TRUNK-5112: Obs Group fix after Patient Merge.

mddubey opened the pull request:
Removing all existing obs with obs-group in encouter recursively and adding new obs with obs-group recursively.

<!--- Provide PR Title above as: 'TRUNK-JiraIssueNumber JiraIssueTitle' -->

## Description
<!--- Describe your changes in detail -->

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue
first -->
<!--- If fixing a bug, there should be an issue describing it with steps to
reproduce -->
<!--- Please link to the issue here: -->
see https://issues.openmrs.org/browse/TRUNK-5112

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that
apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to
help! -->
- [x] My pull request only contains one single commit.
- [x] My pull request is based on the latest master branch
  `git pull --rebase upstream master`.
- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.
- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

dkayiwa commented on api/src/main/java/org/openmrs/api/impl/EncounterServiceImpl.java:
```diff
@@ -217,17 +219,50 @@ public Encounter saveEncounter(Encounter encounter) throws APIException {
-		
+
+		// remove all obs in toRemove recursively from encounter
+		Set<Obs> seenIt = new LinkedHashSet<>();
```
What does seenIt mean?

mddubey replied:
This is just to make sure If there is an obs is added as it's own group member then I should not process that again. Actually I had no such scenarios in my mind(not sure if it is possible as well). I saw a piece of code in `Encounter.java#addObs()`. Then I thought this might be a scenario and can be added as an extra check.

mddubey replied:
I am done with other two things, waiting for a reply from you. Wanted to know if the above check is necessary.

mddubey replied:
One more question, how can I check the Codacy and CodeCoverage things locally? So I can fix them before pushing it.

dkayiwa replied:
If you are not sure whether it will happen or not, then for now, you can start simple by assuming it will not happen. 😊

dkayiwa replied:
@teleivo the codacy question over to you. ☺️

dkayiwa commented on api/src/main/java/org/openmrs/api/impl/EncounterServiceImpl.java:
```diff
@@ -217,17 +219,50 @@ public Encounter saveEncounter(Encounter encounter) throws APIException {
+		return encounter;
+	}
+
+	private void removeAllWithGroupMembers(Encounter encounter, List<Obs> toRemove, Set<Obs> seenIt) {
```
Can we add some javadoc descriptions to these new methods?

dkayiwa commented on api/src/main/java/org/openmrs/api/impl/EncounterServiceImpl.java:
```diff
@@ -217,17 +219,50 @@ public Encounter saveEncounter(Encounter encounter) throws APIException {
+			//prevent infinite recursion if an obs is its own group member
 			encounter.removeObs(o);
+			if (CollectionUtils.isNotEmpty(o.getGroupMembers(true))) {
+				removeAllWithGroupMembers(encounter, new ArrayList<>(o.getGroupMembers(true)), seenIt);
```
Could we assign the new array list to a local variable with a name documenting what it is?

teleivo commented on api/src/main/java/org/openmrs/api/impl/EncounterServiceImpl.java:
```diff
@@ -217,17 +218,51 @@ public Encounter saveEncounter(Encounter encounter) throws APIException {
-		for (Obs o : toRemove) {
+
+		// remove all obs in toRemove recursively from encounter
+		removeAllWithGroupMembers(encounter, toRemove);
```
@mddubey please dont comment your code. you write one line of comment for one line of code. your method is private, you can easily give it a more descriptive name if you feel the need to explain what it does. the method name/signature gives you a chance to explain.

teleivo commented on api/src/test/java/org/openmrs/api/PatientServiceTest.java:
```diff
@@ -3406,4 +3408,33 @@ public void mergePatients_shouldNotFailIfMultiplePatientsHaveActiveOrderOfDiffer
+	 * @see PatientService#mergePatients(org.openmrs.Patient, org.openmrs.Patient)
+	 */
+	@Test
+	@Verifies(value = "Should move all the obs with same hierarchy for the patient after merge.", method = "mergePatient(Patient)")
```
@mddubey this is the old unit test style, see tests before yours. delete the verifies annotation above the method signature. not in the javadoc though.

mddubey replied:
Yeah, I was confused between both, I saw few tests using `@Verifies` annotation, and few javadoc style. That's why I used both, I will delete one of them. I will do the change you mentioned in first comment.

mddubey commented:
@tadeitto done with the changes. One thing I am not getting, the code-coverage is changing Even I don't change anything specifically. E.g.:- I just removed the comments and changed the name of methods it got changed. How actually we calculate that ? I am not sure.

dkayiwa commented on api/src/main/java/org/openmrs/api/impl/EncounterServiceImpl.java:
```diff
@@ -217,17 +218,48 @@ public Encounter saveEncounter(Encounter encounter) throws APIException {
+	 * @param encounter the encounter from which the obs will be removed
+	 * @param obsToRemove list of obs need to be removed recursively
+	 */
+	private void removeObsWithGroupMembers(Encounter encounter, List<Obs> obsToRemove) {
```
The method name seems to suggest removing obs that have (with) group members.
Wouldn't removeObsAndGroupMembers be less confusing?

mddubey replied:
Okay, I will change this one. I was in an impression since I am passing a `list` of obs it should be fine.

dkayiwa commented on api/src/main/java/org/openmrs/api/impl/EncounterServiceImpl.java:
```diff
@@ -217,17 +218,48 @@ public Encounter saveEncounter(Encounter encounter) throws APIException {
+	 * @param encounter the encounter to which the obs will be added
+	 * @param obsToAdd list of obs need to be added recursively
+	 */
+	private void addObsWithGroupMembers(Encounter encounter, List<Obs> obsToAdd) {
```
Same as above regarding the method name.

dkayiwa commented on api/src/main/java/org/openmrs/api/impl/EncounterServiceImpl.java:
```diff
@@ -217,17 +218,48 @@ public Encounter saveEncounter(Encounter encounter) throws APIException {
+	 * This method will add list of obs recursively to encounter
+	 *
+	 * @param encounter the encounter to which the obs will be added
+	 * @param obsToAdd list of obs need to be added recursively
```
Shouldn't we have "that" between "obs" and "need"?

mddubey replied:
I am done with above two changes as well.

dkayiwa commented on api/src/main/java/org/openmrs/api/impl/EncounterServiceImpl.java:
```diff
@@ -217,17 +218,48 @@ public Encounter saveEncounter(Encounter encounter) throws APIException {
+	 * This method will remove list of obs recursively from encounter
+	 *
+	 * @param encounter the encounter from which the obs will be removed
+	 * @param obsToRemove list of obs need to be removed recursively
```
Shouldn't we have "that" between "obs" and "need"?

dkayiwa commented on api/src/main/java/org/openmrs/api/impl/EncounterServiceImpl.java:
```diff
@@ -210,22 +211,51 @@ public Encounter saveEncounter(Encounter encounter) throws APIException {
 			encounter.removeObs(o);
+			Set<Obs> groupMembers = o.getGroupMembers(true);
+			if (CollectionUtils.isNotEmpty(groupMembers)) {
+				ArrayList<Obs> membersToRemove = new ArrayList<>(groupMembers);
```
Could you remind me as to why we instantiate a new collection on the line above?

mddubey replied:
Well, here it is.
In `EncounterServiceImpl#saveEncounter():201` :-  
- We save the existing obs and add to a list named `toRemove`.
- We get a new obs after saving the existing one we add it to a list named `toAdd`.
- Other than this we get the existing obs from the database and add it to a list named `toAdd`.

Since I am removing/adding these obs recursively, I am passing the list and the encounter as a parameter to remove/add method. These are two private methods introduced which is supposed to iterate over list of obs and remove/add from encounter recursively. For recursion I need to do the same for all the group members of each obs in that list, in an iterator, I am calling `obs.getGroupMembers` which I need to pass as other members to add/remove. Since this call returns a `set` and the method accepts a `list`, I am wrapping it into a list and passing it.

Hope that answers your question.

dkayiwa replied:
Can't you just change the method signature from List to Collection?

mddubey replied:
Hey. done with this change as well. This was something stupid which I could have thought of.

dkayiwa commented on api/src/main/java/org/openmrs/api/impl/EncounterServiceImpl.java:
```diff
@@ -210,22 +211,51 @@ public Encounter saveEncounter(Encounter encounter) throws APIException {
 			encounter.addObs(o);
+			Set<Obs> groupMembers = o.getGroupMembers(true);
+			if (CollectionUtils.isNotEmpty(groupMembers)) {
+				ArrayList<Obs> membersToAdd = new ArrayList<>(groupMembers);
```
Could you remind me as to why we instantiate a new collection on the line above?
